csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -215,6 +217,732 @@ class CircularBufferLoopCloner : public kir::IrVisitor {
+      const std::unordered_set<Expr*>& exclude = {}) {
+    TmaCircularBufferLoopCloner cloner(
+        circular_buffer_loop, circular_buffer_load_exprs, loop_type, exclude);
+    cloner.duplicate();
+    return cloner.cloned_top_level_loop_;
+  }
+
+ private:
+  TmaCircularBufferLoopCloner(
+      ForLoop* circular_buffer_loop,
+      const std::vector<Expr*>& circular_buffer_load_exprs,
+      CircularBufferLoopStage loop_type,
+      const std::unordered_set<Expr*>& exclude)
+      : CircularBufferLoopCloner(
+            circular_buffer_loop,
+            circular_buffer_load_exprs,
+            loop_type,
+            exclude) {}
+
+  // For CircularBufferLoopCloner, all load operations are synchronous, so they
+  // are added sequentially. For TmaCircularBufferLoop, we have an mbarrier for
+  // each Tensorview and each circular buffer stage, but not for each individual
+  // TMA load operation. If there are serial IterDomains to the right of the
+  // computeAt position, nvfuser will generate a for-loop to launch multiple TMA
+  // load operations.
+  //
+  // When we encounter a CpAsyncBulk load expression, we create a mbarrier_wait
+  // for the main and epilogue loops and a arriveExpectTx for prologue and main
+  // loops. The expected_tx for arriveExpectTx is the cumulative transaction
+  // size for all TMA load operations for the TensorView. Next, we generate the
+  // nested for-loops for the serial IterDomains, but do not add them to the
+  // cloned circular buffer loop immediately. Once the cloned circular buffer
+  // loop is the only loop in the stack, add the arriveExpectTx and arrive
+  // expressions, then the nested for-loop structure calling the TMA load
+  // operations, and finally the mbarrier_wait.
+  void processForLoop(ForLoop* cloned_loop) final {
+    // Skip if there is not an active for-loop structure
+    if (for_loop_stack_.empty()) {
+      return;
+    }
+
+    if (!cloned_loop->body().empty()) {
+      // mbarrier_arrive_tx_ is active when we encounter a cpAsyncBulk load
+      // operation on a circular buffer TensorView in IrVisitor. A single
+      // mbarrier_arrive_tx is active for each TensorView.
+      if (mbarrier_arrive_tx_ == nullptr || for_loop_stack_.size() > 1) {
+        // Add cloned for_loop when mbarrier_arrive_tx_ is not active or
+        // we are within a nested for-loop structure
+        for_loop_stack_.back()->body().push_back(cloned_loop);
+      } else {
+        // mbarrier::arriveExpectTx and TMA load operations occur in prologue
+        // and main loops.
+        NVF_ERROR(for_loop_stack_.front() == cloned_top_level_loop_);
+        addTmaLoadBlock(cloned_loop);
+      }
+    }
+
+    // mbarrier::wait occurs in Main and Epilogue loops.
+    if (mbarrier_wait_ != nullptr && for_loop_stack_.size() == 1) {
+      NVF_ERROR(for_loop_stack_.back() == cloned_top_level_loop_);
+      cloned_top_level_loop_->body().push_back(mbarrier_wait_);
```

Comment:
For this case, what is `cloned_loop`? Why don't we do `for_loop_stack_.back()->body().push_back(cloned_loop);`?

Reply:
Also, why not append the mbarrier_wait to the body in the function where they are created?

Reply:
`cloned_loop` is the `cloned_circular_buffered_loop`. 

`for_loop_stack_.back() == cloned_top_level_loop_ == cloned_circular_buffer_loop` when `for_loop_stack_.size() == 1`.

> Also, why not append the mbarrier_wait to the body in the function where they are created?

If you're creating a for-loop that launches multiple tma loads, why would you put the `mbarrier_wait` there?

```cpp
if(elect) {
 arriveExpectTx
 for (...) {
  cpAsyncBulk
  mbarrier_wait
 }
} else {
  arrive
}

Shouldn't you add the `mbarrier_wait` when you're at the `cloned_circular_buffer_loop`?